internal/controller/checkpointrestoreoperator_controller.go

```diff
@@ -214,47 +280,180 @@ func getCheckpointArchiveInformation(log logr.Logger, checkpointPath string) (*c
 	return details, nil
 }
 
+func applyPolicies(log logr.Logger, details *checkpointDetails) {
+	policyMutex.Lock()
+	defer policyMutex.Unlock()
+
+	if policy := findContainerPolicy(details); policy != nil {
+		log.Info("Applying container-level policy", "container", details.container)
+		handleCheckpointsForLevel(log, details, "container", int(*policy.MaxCheckpoints))
+	} else if policy := findPodPolicy(details); policy != nil {
+		log.Info("Applying pod-level policy", "pod", details.pod)
+		handleCheckpointsForLevel(log, details, "pod", int(*policy.MaxCheckpoints))
+	} else if policy := findNamespacePolicy(details); policy != nil {
+		log.Info("Applying namespace-level policy", "namespace", details.namespace)
+		handleCheckpointsForLevel(log, details, "namespace", int(*policy.MaxCheckpoints))
+	} else {
+		// Apply global policies if no specific policy found
+		log.Info("Applying global container-level policy", "container", details.container)
+		handleCheckpointsForLevel(log, details, "container", maxCheckpointsPerContainer)
+
+		log.Info("Applying global pod-level policy", "pod", details.pod)
+		handleCheckpointsForLevel(log, details, "pod", maxCheckpointsPerPod)
+
+		log.Info("Applying global namespace-level policy", "namespace", details.namespace)
+		handleCheckpointsForLevel(log, details, "namespace", maxCheckpointsPerNamespace)
+	}
+}
+
+func findContainerPolicy(details *checkpointDetails) *criuorgv1.ContainerPolicySpec {
+	for _, policy := range containerPolicies {
+		if policy.Namespace == details.namespace && policy.Pod == details.pod && policy.Container == details.container {
+			return &policy
+		}
+	}
+	return nil
+}
+
+func findPodPolicy(details *checkpointDetails) *criuorgv1.PodPolicySpec {
+	for _, policy := range podPolicies {
+		if policy.Namespace == details.namespace && policy.Pod == details.pod {
+			return &policy
+		}
+	}
+	return nil
+}
+
+func findNamespacePolicy(details *checkpointDetails) *criuorgv1.NamespacePolicySpec {
+	for _, policy := range namespacePolicies {
+		if policy.Namespace == details.namespace {
+			return &policy
+		}
+	}
+	return nil
+}
+
+func applyPoliciesImmediately(log logr.Logger, checkpointDirectory string) {
+	log.Info("Applying policies immediately")
+
+	checkpointFiles, err := filepath.Glob(filepath.Join(checkpointDirectory, "checkpoint-*_*-*-*.tar"))
+	if err != nil {
+		log.Error(err, "Failed to list checkpoint files")
+		return
+	}
+
+	if len(checkpointFiles) == 0 {
+		log.Info("No checkpoint files found")
+		files, err := os.ReadDir(checkpointDirectory)
+		if err != nil {
+			log.Error(err, "Failed to list contents of directory")
+		} else {
+			log.Info("Contents of the directory:")
+			for _, file := range files {
```

Comment:
I'm not sure if this would be a good idea. Printing the file name of all checkpoints on a large scale system might be very expensive.

Reply:
I think I did this for debugging purpose, and forgot to remove it later.